packages/manager/apps/pci-ai-endpoints/src/hooks/metric/useGenerateMetricData.hook.tsx

```diff
@@ -0,0 +1,179 @@
+import { useMemo } from 'react';
+import {
+  format,
+  parse,
+  isSameDay,
+  differenceInMonths,
+  startOfWeek,
+  addDays,
+  addWeeks,
+  addMonths,
+  startOfMonth,
+  isAfter,
+} from 'date-fns';
+import { HostMetric } from '@/types/cloud/project/database/metric';
+
+// Function to format a Date object into "DD/MM/YYYY" format
+const formatDate = (date: Date): string => format(date, 'dd/MM/yyyy');
+
+// Function to parse a "DD/MM/YYYY" formatted string into a Date object
+const parseDate = (dateString: string | Date): Date => {
+  if (typeof dateString === 'string') {
+    return parse(dateString, 'dd/MM/yyyy', new Date());
+  }
+  return dateString;
+};
+
+// Define a mapping for units to human-readable labels
+// Define a type for the unit labels
+type UnitLabels = 'input' | 'output' | 'num_requests';
+
+// Map the units to their expected labels
+const unitLabels: Record<UnitLabels, string> = {
+  input: 'input_tokens',
+  output: 'output_tokens',
+  num_requests: 'num_requests',
+};
+
+// Hook to generate labels based on the selected date range
+const useGenerateLabels = (start: string | Date, end: string | Date) => {
```

Comment:
Can you wrap big calls like every call on date inside useMemo to avoid re-renders